std/meta/index.zig

```diff
@@ -0,0 +1,691 @@
+        TypeId.Union => |info| info.fields,
+        TypeId.Enum => |info| info.fields,
+        TypeId.ErrorSet => |info| info.errors,
+        else => @compileError("Expected struct, union, error set or enum type, found '"
+            ++ @typeName(T) ++ "'"),
+    };
+}
+
+test "std.meta.fields" {
+    const E1 = enum.{
+        A,
+    };
+    const E2 = error.{A};
+    const S1 = struct.{
+        a: u8,
+    };
+    const U1 = union.{
+        a: u8,
+    };
+
+    const e1f = comptime fields(E1);
+    const e2f = comptime fields(E2);
+    const sf = comptime fields(S1);
+    const uf = comptime fields(U1);
+
+    debug.assert(e1f.len == 1);
+    debug.assert(e2f.len == 1);
+    debug.assert(sf.len == 1);
+    debug.assert(uf.len == 1);
+    debug.assert(mem.eql(u8, e1f[0].name, "A"));
+    debug.assert(mem.eql(u8, e2f[0].name, "A"));
+    debug.assert(mem.eql(u8, sf[0].name, "a"));
+    debug.assert(mem.eql(u8, uf[0].name, "a"));
+    debug.assert(comptime sf[0].field_type == u8);
+    debug.assert(comptime uf[0].field_type == u8);
+}
+
+pub fn fieldInfo(comptime T: type, comptime field_name: []const u8) switch (@typeInfo(T)) {
+    TypeId.Struct => TypeInfo.StructField,
+    TypeId.Union => TypeInfo.UnionField,
+    TypeId.ErrorSet => TypeInfo.Error,
+    TypeId.Enum => TypeInfo.EnumField,
+    else => @compileError("Expected struct, union, error set or enum type, found '" 
+        ++ @typeName(T) ++ "'"),
+} {
+    inline for (comptime fields(T)) |field| {
+        if (comptime mem.eql(u8, field.name, field_name))
+            return field;
+    }
+
+    @compileError("'" ++ @typeName(T) ++ "' has no field '" ++ field_name ++ "'");
+}
+
+test "std.meta.fieldInfo" {
+    const E1 = enum.{
+        A,
+    };
+    const E2 = error.{A};
+    const S1 = struct.{
+        a: u8,
+    };
+    const U1 = union.{
+        a: u8,
+    };
+
+    const e1f = comptime fieldInfo(E1, "A");
+    const e2f = comptime fieldInfo(E2, "A");
+    const sf = comptime fieldInfo(S1, "a");
+    const uf = comptime fieldInfo(U1, "a");
+
+    debug.assert(mem.eql(u8, e1f.name, "A"));
+    debug.assert(mem.eql(u8, e2f.name, "A"));
+    debug.assert(mem.eql(u8, sf.name, "a"));
+    debug.assert(mem.eql(u8, uf.name, "a"));
+    debug.assert(comptime sf.field_type == u8);
+    debug.assert(comptime uf.field_type == u8);
+}
+
+pub fn TagType(comptime T: type) type {
+    return switch (@typeInfo(T)) {
+        TypeId.Enum => |info| info.tag_type,
+        TypeId.Union => |info| if(info.tag_type) |Tag| Tag else null,
+        else => @compileError("expected enum or union type, found '" ++ @typeName(T) ++ "'"),
+    };
+}
+
+test "std.meta.TagType" {
+    const E = enum(u8).{
+        C = 33,
+        D,
+    };
+    const U = union(E).{
+        C: u8,
+        D: u16,
+    };
+
+    debug.assert(TagType(E) == u8);
+    debug.assert(TagType(U) == E);
+}
+
+
+
+///Returns the active tag of a tagged union
+pub fn activeTag(u: var) @TagType(UnwrapContainer(@typeOf(u)))
```

Comment:
I don't think this function should take pointers and deref it. It should just take the value, and the user is responsible for the deref.

Reply:
That makes sense, given that the compiler should ensure that no copies of large structs occur.